pygfx/objects/_events.py

```diff
@@ -0,0 +1,358 @@
+from collections import defaultdict
+from enum import Enum
+from time import perf_counter_ns
+from typing import Union
+from weakref import ref, WeakValueDictionary
+
+
+CLICK_DEBOUNCE = 500  # in milliseconds
+
+
+class EventType(str, Enum):
+    # Keyboard
+    KEY_DOWN = "key_down"
+    KEY_UP = "key_up"
+    # Pointer
+    POINTER_DOWN = "pointer_down"
+    POINTER_MOVE = "pointer_move"
+    POINTER_UP = "pointer_up"
+    CLICK = "click"
+    DOUBLE_CLICK = "double_click"
+    # Wheel
+    WHEEL = "wheel"
+    # Window
+    CLOSE = "close"
+    RESIZE = "resize"
+
+
+class Event:
+    """Event base class for creating events.
+
+    If a target is set, an event can bubble up through a hierarchy
+    of targets, connected through a ``parent`` property.
+    To prevent an event from bubbling up, use ``stop_propagation``.
+
+    It is also possible to cancel events, which will stop any further
+    handling of the event (also by the same target).
+
+    Custom events can have any fields. Any unknown ``kwargs`` will be
+    captured and can be later retrieved with the square bracket notation.
+    """
+
+    def __init__(
+        self,
+        type: Union[str, EventType],
+        *,
+        bubbles=True,
+        target: "EventTarget" = None,
+        time_stamp: float = None,
+        cancelled: bool = False,
+        **kwargs,
+    ):
+        self._type = type
+        # Using perf_counter_ns instead of perf_counter
+        # should give us a bit more accuracy
+        self._time_stamp = time_stamp or perf_counter_ns() / 1000000
```

Comment:
I wouldn't have guessed that it mattered for precision.

> Note that the standard on the web is an integer in milliseconds.

Yes, but this is Python :)  or do you mean that since this system is so similar to web events, that it might make sense to follow the web here too? (which is a valid argument too)

> I don't mind which choice we make ultimately.

Indeed, it is not really clear which is more natural, and it also does not matter so much. 🤷